File: kdeui/src/common/coordinates.cpp

/**************************************************************************
 *                                                                        *
 *  Regina - A Normal Surface Theory Calculator                           *
 *  KDE User Interface                                                    *
 *                                                                        *
 *  Copyright (c) 1999-2009, Ben Burton                                   *
 *  For further details contact Ben Burton (bab@debian.org).              *
 *                                                                        *
 *  This program is free software; you can redistribute it and/or         *
 *  modify it under the terms of the GNU General Public License as        *
 *  published by the Free Software Foundation; either version 2 of the    *
 *  License, or (at your option) any later version.                       *
 *                                                                        *
 *  This program is distributed in the hope that it will be useful, but   *
 *  WITHOUT ANY WARRANTY; without even the implied warranty of            *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU     *
 *  General Public License for more details.                              *
 *                                                                        *
 *  You should have received a copy of the GNU General Public             *
 *  License along with this program; if not, write to the Free            *
 *  Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,       *
 *  MA 02110-1301, USA.                                                   *
 *                                                                        *
 **************************************************************************/

/* end stub */

#include "surfaces/nnormalsurfacelist.h"
#include "triangulation/ntriangulation.h"

#include <klocale.h>

using regina::NNormalSurfaceList;

namespace Coordinates {
    QString name(int coordSystem, bool capitalise) {
        if (capitalise) {
            if (coordSystem == NNormalSurfaceList::STANDARD)
                return i18n("Standard normal (tri-quad)");
            if (coordSystem == NNormalSurfaceList::AN_STANDARD)
                return i18n("Standard almost normal (tri-quad-oct)");
            if (coordSystem == NNormalSurfaceList::AN_LEGACY)
                return i18n("Legacy almost normal (pruned tri-quad-oct)");
            if (coordSystem == NNormalSurfaceList::QUAD)
                return i18n("Quad normal");
            if (coordSystem == NNormalSurfaceList::AN_QUAD_OCT)
                return i18n("Quad-oct almost normal");
            if (coordSystem == NNormalSurfaceList::EDGE_WEIGHT)
                return i18n("Edge weight");
            if (coordSystem == NNormalSurfaceList::FACE_ARCS)
                return i18n("Face arc");
            if (coordSystem == NNormalSurfaceList::ORIENTED)
                return i18n("Transversely oriented normal (tri-quad)");
            if (coordSystem == NNormalSurfaceList::ORIENTED_QUAD)
                return i18n("Transversely oriented quad normal (quad)");
            return i18n("Unknown");
        } else {
            if (coordSystem == NNormalSurfaceList::STANDARD)
                return i18n("standard normal (tri-quad)");
            if (coordSystem == NNormalSurfaceList::AN_STANDARD)
                return i18n("standard almost normal (tri-quad-oct)");
            if (coordSystem == NNormalSurfaceList::AN_LEGACY)
                return i18n("legacy almost normal (pruned tri-quad-oct)");
            if (coordSystem == NNormalSurfaceList::QUAD)
                return i18n("quad normal");
            if (coordSystem == NNormalSurfaceList::AN_QUAD_OCT)
                return i18n("quad-oct almost normal");
            if (coordSystem == NNormalSurfaceList::EDGE_WEIGHT)
                return i18n("edge weight");
            if (coordSystem == NNormalSurfaceList::FACE_ARCS)
                return i18n("face arc");
            if (coordSystem == NNormalSurfaceList::ORIENTED)
                return i18n("transversely oriented normal (tri-quad)");
            if (coordSystem == NNormalSurfaceList::ORIENTED_QUAD)
                return i18n("transversely oriented quad normal (quad)");
            return i18n("unknown");
        }
    }

    bool generatesAlmostNormal(int coordSystem) {
        return (coordSystem == NNormalSurfaceList::AN_STANDARD ||
                coordSystem == NNormalSurfaceList::AN_LEGACY ||
                coordSystem == NNormalSurfaceList::AN_QUAD_OCT);
    }

    unsigned long numColumns(int coordSystem, regina::NTriangulation* tri) {
        if (coordSystem == NNormalSurfaceList::STANDARD)
            return tri->getNumberOfTetrahedra() * 7;
        else if (coordSystem == NNormalSurfaceList::AN_STANDARD)
            return tri->getNumberOfTetrahedra() * 10;
        else if (coordSystem == NNormalSurfaceList::AN_LEGACY)
            return tri->getNumberOfTetrahedra() * 10;
        else if (coordSystem == NNormalSurfaceList::QUAD)
            return tri->getNumberOfTetrahedra() * 3;
        else if (coordSystem == NNormalSurfaceList::AN_QUAD_OCT)
            return tri->getNumberOfTetrahedra() * 6;
        else if (coordSystem == NNormalSurfaceList::EDGE_WEIGHT)
            return tri->getNumberOfEdges();
        else if (coordSystem == NNormalSurfaceList::FACE_ARCS)
            return tri->getNumberOfFaces() * 3;
        else if (coordSystem == NNormalSurfaceList::ORIENTED)
            return tri->getNumberOfTetrahedra() * 14;
        else if (coordSystem == NNormalSurfaceList::ORIENTED)
            return tri->getNumberOfTetrahedra() * 8;
        else
            return 0;
    }

    QString columnName(int coordSystem, unsigned long whichCoord,
            regina::NTriangulation* tri) {
        if (coordSystem == NNormalSurfaceList::STANDARD) {
            if (whichCoord % 7 < 4)
                return QString("%1: %2").arg(whichCoord / 7).
                    arg(whichCoord % 7);
            else
                return QString("%1: %2").arg(whichCoord / 7).
                    arg(regina::vertexSplitString[(whichCoord % 7) - 4]);
        } else if (coordSystem == NNormalSurfaceList::AN_STANDARD ||
                coordSystem == NNormalSurfaceList::AN_LEGACY) {
            if (whichCoord % 10 < 4)
                return i18n("T%1: %2").arg(whichCoord / 10).
                    arg(whichCoord % 10);
            else if (whichCoord % 10 < 7)
                return i18n("Q%1: %2").arg(whichCoord / 10).
                    arg(regina::vertexSplitString[(whichCoord % 10) - 4]);
            else
                return i18n("K%1: %2").arg(whichCoord / 10).
                    arg(regina::vertexSplitString[(whichCoord % 10) - 7]);
        } else if (coordSystem == NNormalSurfaceList::QUAD) {
            return QString("%1: %2").arg(whichCoord / 3).
                arg(regina::vertexSplitString[whichCoord % 3]);
        } else if (coordSystem == NNormalSurfaceList::AN_QUAD_OCT) {
            if (whichCoord % 6 < 3)
                return i18n("Q%1: %2").arg(whichCoord / 6).
                    arg(regina::vertexSplitString[whichCoord % 6]);
            else
                return i18n("K%1: %2").arg(whichCoord / 6).
                    arg(regina::vertexSplitString[(whichCoord % 6) - 3]);
        } else if (coordSystem == NNormalSurfaceList::EDGE_WEIGHT) {
            if (! (tri && tri->getEdge(whichCoord)->isBoundary()))
                return QString::number(whichCoord);
            else
                return i18n("%1 [B]").arg(whichCoord);
        } else if (coordSystem == NNormalSurfaceList::FACE_ARCS) {
            return QString("%1: %2").arg(whichCoord / 3).arg(whichCoord % 3);
        } else if (coordSystem == NNormalSurfaceList::ORIENTED) {
            QString orientation = "+";
            unsigned long coord = whichCoord;
            if (whichCoord % 2 == 1) {
                coord--;
                orientation = "-";
            }
            coord = coord / 2;
            if (coord % 7 < 4)
                return QString("%1: %2 (%3)").arg(coord / 7).
                    arg(coord % 7).
                    arg(orientation);
            else
                return QString("%1: %2 (%3)").arg(coord / 7).
                    arg(regina::vertexSplitString[(coord % 7) - 4]).
                    arg(orientation);
        } else if (coordSystem == NNormalSurfaceList::ORIENTED_QUAD) {
            QString orientation = "+";
            unsigned long coord = whichCoord;
            if (whichCoord % 2 == 1) {
                coord--;
                orientation = "-";
            }
            coord = coord / 2;
            return QString("%1: %2 (%3)").arg(coord / 3).
                arg(regina::vertexSplitString[coord % 3]).
                arg(orientation);
        }

        return i18n("Unknown");
    }

    QString columnDesc(int coordSystem, unsigned long whichCoord,
            regina::NTriangulation* tri) {
        if (coordSystem == NNormalSurfaceList::STANDARD) {
            if (whichCoord % 7 < 4)
                return i18n("Tetrahedron %1, triangle about vertex %2").
                    arg(whichCoord / 7).arg(whichCoord % 7);
            else
                return i18n("Tetrahedron %1, quad splitting vertices %2").
                    arg(whichCoord / 7).
                    arg(regina::vertexSplitString[(whichCoord % 7) - 4]);
        } else if (coordSystem == NNormalSurfaceList::AN_STANDARD ||
                coordSystem == NNormalSurfaceList::AN_LEGACY) {
            if (whichCoord % 10 < 4)
                return i18n("Tetrahedron %1, triangle about vertex %2").
                    arg(whichCoord / 10).arg(whichCoord % 10);
            else if (whichCoord % 10 < 7)
                return i18n("Tetrahedron %1, quad splitting vertices %2").
                    arg(whichCoord / 10).
                    arg(regina::vertexSplitString[(whichCoord % 10) - 4]);
            else
                return i18n("Tetrahedron %1, oct partitioning vertices %2").
                    arg(whichCoord / 10).
                    arg(regina::vertexSplitString[(whichCoord % 10) - 7]);
        } else if (coordSystem == NNormalSurfaceList::QUAD) {
            return i18n("Tetrahedron %1, quad splitting vertices %2").
                arg(whichCoord / 3).
                arg(regina::vertexSplitString[whichCoord % 3]);
        } else if (coordSystem == NNormalSurfaceList::AN_QUAD_OCT) {
            if (whichCoord % 6 < 3)
                return i18n("Tetrahedron %1, quad splitting vertices %2").
                    arg(whichCoord / 6).
                    arg(regina::vertexSplitString[whichCoord % 6]);
            else
                return i18n("Tetrahedron %1, oct partitioning vertices %2").
                    arg(whichCoord / 6).
                    arg(regina::vertexSplitString[(whichCoord % 6) - 3]);
        } else if (coordSystem == NNormalSurfaceList::EDGE_WEIGHT) {
            if (tri) {
                if (tri->getEdge(whichCoord)->isBoundary())
                    return i18n("Weight of (boundary) edge %1").arg(whichCoord);
                else
                    return i18n("Weight of (internal) edge %1").arg(whichCoord);
            } else
                return i18n("Weight of edge %1").arg(whichCoord);
        } else if (coordSystem == NNormalSurfaceList::FACE_ARCS) {
            return i18n("Arcs on face %1 crossing face vertex %2").
                arg(whichCoord / 3).arg(whichCoord % 3);
        } else if (coordSystem == NNormalSurfaceList::ORIENTED) {
            QString orientation = "+";
            unsigned long coord = whichCoord;
            if (whichCoord % 2 == 1) {
                coord--;
                orientation = "-";
            }
            coord = coord / 2;
            if (coord % 7 < 4)
                return i18n("Tetrahedron %1, triangle about vertex %2, "
                    "orientation %3").
                    arg(coord / 7).arg(coord % 7).
                    arg(orientation);
            else
                return i18n("Tetrahedron %1, quad splitting vertices %2, "
                    "orientation %3").
                    arg(coord / 7).
                    arg(regina::vertexSplitString[(coord % 7) - 4]).
                    arg(orientation);
        } else if (coordSystem == NNormalSurfaceList::ORIENTED_QUAD) {
            QString orientation = "+";
            unsigned long coord = whichCoord;
            if (whichCoord % 2 == 1) {
                coord--;
                orientation = "-";
            }
            coord = coord / 2;
            return i18n("Tetrahedron %1, quad splitting vertices %2, "
                "orientation %3").
                arg(coord / 3).
                arg(regina::vertexSplitString[coord % 3]).
                arg(orientation);
        }

        return i18n("This coordinate system is not known");
    }

    regina::NLargeInteger getCoordinate(int coordSystem,
            const regina::NNormalSurface& surface, unsigned long whichCoord) {
        if (coordSystem == NNormalSurfaceList::STANDARD) {
            if (whichCoord % 7 < 4)
                return surface.getTriangleCoord(
                    whichCoord / 7, whichCoord % 7);
            else
                return surface.getQuadCoord(
                    whichCoord / 7, (whichCoord % 7) - 4);
        } else if (coordSystem == NNormalSurfaceList::AN_STANDARD ||
                coordSystem == NNormalSurfaceList::AN_LEGACY) {
            if (whichCoord % 10 < 4)
                return surface.getTriangleCoord(
                    whichCoord / 10, whichCoord % 10);
            else if (whichCoord % 10 < 7)
                return surface.getQuadCoord(
                    whichCoord / 10, (whichCoord % 10) - 4);
            else
                return surface.getOctCoord(
                    whichCoord / 10, (whichCoord % 10) - 7);
        } else if (coordSystem == NNormalSurfaceList::QUAD) {
            return surface.getQuadCoord(whichCoord / 3, whichCoord % 3);
        } else if (coordSystem == NNormalSurfaceList::AN_QUAD_OCT) {
            if (whichCoord % 6 < 3)
                return surface.getQuadCoord(whichCoord / 6, whichCoord % 6);
            else
                return surface.getOctCoord(
                    whichCoord / 6, (whichCoord % 6) - 3);
        } else if (coordSystem == NNormalSurfaceList::EDGE_WEIGHT) {
            return surface.getEdgeWeight(whichCoord);
        } else if (coordSystem == NNormalSurfaceList::FACE_ARCS) {
            return surface.getFaceArcs(whichCoord / 3, whichCoord % 3);
        } else if (coordSystem == NNormalSurfaceList::ORIENTED) {
            bool orientation = true;
            unsigned long coord = whichCoord;
            if (whichCoord % 2 == 1) {
                coord--;
                orientation = false;
            }
            coord = coord / 2;
            if (coord % 7 < 4)
                return surface.getTriangleCoord(
                    coord / 7, coord % 7,orientation);
            else
                return surface.getQuadCoord(
                    coord / 7, (coord % 7) - 4,orientation);
        } else if (coordSystem == NNormalSurfaceList::ORIENTED_QUAD) {
            bool orientation = true;
            unsigned long coord = whichCoord;
            if (whichCoord % 2 == 1) {
                coord--;
                orientation = false;
            }
            coord = coord / 2;
            return surface.getQuadCoord(whichCoord / 3, coord % 3, orientation);
        }

        return (long)0;
    }
};

File: javaui/normal/packetui/surfaces/Coordinates.java

/**************************************************************************
 *                                                                        *
 *  Regina - A Normal Surface Theory Calculator                           *
 *  Java User Interface                                                   *
 *                                                                        *
 *  Copyright (c) 1999-2002, Ben Burton                                   *
 *  For further details contact Ben Burton (bab@debian.org).              *
 *                                                                        *
 *  This program is free software; you can redistribute it and/or         *
 *  modify it under the terms of the GNU General Public License as        *
 *  published by the Free Software Foundation; either version 2 of the    *
 *  License, or (at your option) any later version.                       *
 *                                                                        *
 *  This program is distributed in the hope that it will be useful, but   *
 *  WITHOUT ANY WARRANTY; without even the implied warranty of            *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU     *
 *  General Public License for more details.                              *
 *                                                                        *
 *  You should have received a copy of the GNU General Public             *
 *  License along with this program; if not, write to the Free            *
 *  Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,        *
 *  MA 02111-1307, USA.                                                   *
 *                                                                        *
 **************************************************************************/

/* end stub */

package normal.packetui.surfaces;

import normal.engine.surfaces.*;
import normal.engine.triangulation.NTriangulation;
import java.math.BigInteger;
import java.util.Vector;

/**
 * Contains routines to handle the variety of available coordinate
 * systems.  Whenever a new flavour of coordinate system is introduced,
 * only this class should need to be edited.
 */
public class Coordinates {
    /**
     * Returns a list of all flavours of coordinate system available for
     * creating normal surface lists.  Each object in the returned
     * vector will be of type <tt>Coordinates</tt> and will be suitable
     * for placing in a combo box or the like.
     *
     * @return a list of coordinate systems.
     */
    public static Vector allCreators() {
        Vector ans = new Vector();
        ans.addElement(new Coordinates(NNormalSurfaceList.STANDARD));
        ans.addElement(new Coordinates(NNormalSurfaceList.AN_STANDARD));
        ans.addElement(new Coordinates(NNormalSurfaceList.QUAD));
        return ans;
    }

    /**
     * Returns a list of all flavours of coordinate system capable of
     * viewing the given normal surface set.  Each object in the returned
     * vector will be of type <tt>Coordinates</tt> and will be suitable
     * for placing in a combo box or the like.
     * <p>
     * If the given set is <tt>null</tt>, an empty vector will be
     * returned.
     *
     * @param set the normal surface set to be viewed.
     * @return a list of coordinate systems.
     */
    public static Vector allViewers(NSurfaceSet set) {
        if (set == null)
            return new Vector();

        boolean almostNormal = set.allowsAlmostNormal();
        Vector ans = new Vector();
        if (almostNormal) {
            ans.addElement(new Coordinates(NNormalSurfaceList.AN_STANDARD));
        } else {
            ans.addElement(new Coordinates(NNormalSurfaceList.STANDARD));
            ans.addElement(new Coordinates(NNormalSurfaceList.QUAD));
        }
        ans.addElement(new Coordinates(NNormalSurfaceList.EDGE_WEIGHT));
        ans.addElement(new Coordinates(NNormalSurfaceList.FACE_ARCS));
        return ans;
    }

    /**
     * Returns the name of the given flavour of coordinate system.
     * The name will begin with a capital and <i>not</i> end with the word
     * <i>coordinates</i> or a period.
     *
     * @param flavour the flavour of coordinate system to use; this must
     * be one of the coordinate system constants defined in
     * <tt>NNormalSurfaceList</tt>.
     * @return the name of the given flavour of coordinate system.
     * @see normal.engine.surfaces.NNormalSurfaceList
     */
    public static String coordinateSystemName(int flavour) {
        switch(flavour) {
            case NNormalSurfaceList.STANDARD:
                return "Standard normal (tri-quad)";
            case NNormalSurfaceList.AN_STANDARD:
                return "Standard almost normal (tri-quad-oct)";

            case NNormalSurfaceList.QUAD:
                return "Quad normal";

            case NNormalSurfaceList.EDGE_WEIGHT:
                return "Edge weight";
            case NNormalSurfaceList.FACE_ARCS:
                return "Face arcs";
        }
        return null;
    }

    /**
     * Returns an abbreviation for the given coordinate in the given
     * flavour of coordinate system.
     *
     * @param flavour the flavour of coordinate system to use; this must
     * be one of the coordinate system constants defined in
     * <tt>NNormalSurfaceList</tt>.
     * @param triang the triangulation in which the normal surface is to
     * reside.
     * @param coord the coordinate to examine; this should be between 0
     * and one less than the total number of coordinates (inclusive).
     * @return an abbreviation for the given coordinate.
     * @see normal.engine.surfaces.NNormalSurfaceList
     */
    public static String getCoordinateAbbr(int flavour,
            NTriangulation triang, int coord) {
        String ans;
        int tmp;
        switch(flavour) {
            case NNormalSurfaceList.STANDARD:
                tmp = coord % 7;
                ans = String.valueOf(coord / 7) + ": ";
                if (tmp < 4)
                    return ans + String.valueOf(tmp);
                else
                    return ans + NNormalSurface.vertexSplitString[tmp - 4];
            case NNormalSurfaceList.AN_STANDARD:
                tmp = coord % 10;
                ans = String.valueOf(coord / 10) + ": ";
                if (tmp < 4)
                    return "T" + ans + String.valueOf(tmp);
                else if (tmp < 7)
                    return "Q" + ans +
                        NNormalSurface.vertexSplitString[tmp - 4];
                else
                    return "K" + ans +
                        NNormalSurface.vertexSplitString[tmp - 7];
            case NNormalSurfaceList.QUAD:
                return String.valueOf(coord / 3) + ": " +
                    NNormalSurface.vertexSplitString[coord % 3];
            case NNormalSurfaceList.EDGE_WEIGHT:
                if (triang.getEdge(coord).isBoundary())
                    return String.valueOf(coord) + " [B]";
                else
                    return String.valueOf(coord);
            case NNormalSurfaceList.FACE_ARCS:
                return String.valueOf(coord / 3) + ": " +
                    String.valueOf(coord % 3);
        }
        return null;
    }

    /**
     * Returns a description of the given coordinate in the given
     * flavour of coordinate system.
     *
     * @param flavour the flavour of coordinate system to use; this must
     * be one of the coordinate system constants defined in
     * <tt>NNormalSurfaceList</tt>.
     * @param triang the triangulation in which the normal surface is to
     * reside.
     * @param coord the coordinate to examine; this should be between 0
     * and one less than the total number of coordinates (inclusive).
     * @return a description of the given coordinate.
     * @see normal.engine.surfaces.NNormalSurfaceList
     */
    public static String getCoordinateDesc(int flavour,
            NTriangulation triang, int coord) {
        String ans;
        int tmp;
        switch(flavour) {
            case NNormalSurfaceList.STANDARD:
                tmp = coord % 7;
                ans = "Tetrahedron " + String.valueOf(coord / 7) + ", ";
                if (tmp < 4)
                    return ans + "triangle about vertex " + String.valueOf(tmp);
                else
                    return ans + "quad splitting vertices " +
                        NNormalSurface.vertexSplitString[tmp - 4];
            case NNormalSurfaceList.AN_STANDARD:
                tmp = coord % 10;
                ans = "Tetrahedron " + String.valueOf(coord / 10) + ", ";
                if (tmp < 4)
                    return ans + "triangle about vertex " + String.valueOf(tmp);
                else if (tmp < 7)
                    return ans + "quad splitting vertices " +
                        NNormalSurface.vertexSplitString[tmp - 4];
                else
                    return ans + "oct partitioning vertices " +
                        NNormalSurface.vertexSplitString[tmp - 7];
            case NNormalSurfaceList.QUAD:
                return "Tetrahedron " + String.valueOf(coord / 3) +
                    ", quad splitting vertices " +
                    NNormalSurface.vertexSplitString[coord % 3];
            case NNormalSurfaceList.EDGE_WEIGHT:
                if (triang.getEdge(coord).isBoundary())
                    return "Weight of (boundary) edge " + String.valueOf(coord);
                else
                    return "Weight of (internal) edge " + String.valueOf(coord);
            case NNormalSurfaceList.FACE_ARCS:
                return "Arcs on face " + String.valueOf(coord / 3) +
                    " crossing face vertex " + String.valueOf(coord % 3);
        }
        return null;
    }

    /**
     * Returns the number of coordinates required for a normal surface
     * in the given triangulation using the given flavour of coordinate
     * system.
     *
     * @param flavour the flavour of coordinate system to use; this must
     * be one of the coordinate system constants defined in
     * <tt>NNormalSurfaceList</tt>.
     * @param triang the triangulation in which the normal surface is to
     * reside.
     * @return the number of coordinate required.
     * @see normal.engine.surfaces.NNormalSurfaceList
     */
    public static int getNumberOfCoordinates(int flavour,
            NTriangulation triang) {
        switch(flavour) {
            case NNormalSurfaceList.STANDARD:
                return 7 * (int)triang.getNumberOfTetrahedra();
            case NNormalSurfaceList.AN_STANDARD:
                return 10 * (int)triang.getNumberOfTetrahedra();
            case NNormalSurfaceList.QUAD:
                return 3 * (int)triang.getNumberOfTetrahedra();
            case NNormalSurfaceList.EDGE_WEIGHT:
                return (int)triang.getNumberOfEdges();
            case NNormalSurfaceList.FACE_ARCS:
                return 3 * (int)triang.getNumberOfFaces();
        }
        return 0;
    }

    /**
     * Returns the requested coordinate for the given normal surface in
     * the given flavour of coordinate system.
     *
     * @param flavour the flavour of coordinate system to use; this must
     * be one of the coordinate system constants defined in
     * <tt>NNormalSurfaceList</tt>.
     * @param surface the normal surface whose coordinates we are examining.
     * @param coord the coordinate to examine; this should be between 0
     * and one less than the total number of coordinates for the given
     * flavour of coordinate system (inclusive).
     * @return the requested coordinate in the given normal surface.
     * @see normal.engine.surfaces.NNormalSurfaceList
     */
    public static BigInteger getCoordinate(int flavour,
            NNormalSurface surface, int coord) {
        int tmp;
        switch(flavour) {
            case NNormalSurfaceList.STANDARD:
                tmp = coord % 7;
                if (tmp < 4)
                    return surface.getTriangleCoord(coord / 7, tmp);
                else
                    return surface.getQuadCoord(coord / 7, tmp - 4);
            case NNormalSurfaceList.AN_STANDARD:
                tmp = coord % 10;
                if (tmp < 4)
                    return surface.getTriangleCoord(coord / 10, tmp);
                else if (tmp < 7)
                    return surface.getQuadCoord(coord / 10, tmp - 4);
                else
                    return surface.getOctCoord(coord / 10, tmp - 7);
            case NNormalSurfaceList.QUAD:
                return surface.getQuadCoord(coord / 3, coord % 3);
            case NNormalSurfaceList.EDGE_WEIGHT:
                return surface.getEdgeWeight(coord);
            case NNormalSurfaceList.FACE_ARCS:
                return surface.getFaceArcs(coord / 3, coord % 3);
        }
        return null;
    }

    /**
     * The constant representing this flavour of coordinate system.
     */
    private int flavour;
    
    /**
     * Creates a new coordinate system flavour object.
     *
     * @param flavour the constant representing this flavour of
     * coordinate system; this should be one of the coordinate system
     * constants defined in <tt>NNormalSurfaceList</tt>.
     * @see normal.engine.surfaces.NNormalSurfaceList
     */
    public Coordinates(int flavour) {
        this.flavour = flavour;
    }

    /**
     * Returns the constant representing this flavour of coordinate
     * system.  This is the constant that should be passed to the
     * <tt>NNormalSurfaceList</tt> constructor.
     *
     * @return the constant representing this coordinate system.
     * @see normal.engine.Engine#newNNormalSurfaceList
     */
    public int getFlavour() {
        return flavour;
    }

    /**
     * Returns the name of this flavour of coordinate system.  This
     * is the string that should be displayed in a combo box or the like.
     *
     * @return the name of this coordinate system.
     */
    public String toString() {
        return coordinateSystemName(flavour);
    }

    /**
     * Determines if this and the given object represent the same
     * flavour of coordinate system.
     *
     * @param other the object to compare with this.
     * @return <tt>true</tt> if and only if this and the given object
     * represent the same flavour of coordinate system.
     */
    public boolean equals(Object other) {
        if (! (other instanceof Coordinates))
            return false;
        return (flavour == ((Coordinates)other).flavour);
    }
}
